lib/remote_node.go

```diff
@@ -0,0 +1,599 @@
+// nodes have sent and received a Version and Verack message. Once the handshake is successful, the RemoteNode will
+// emit a MsgDeSoPeerHandshakeComplete control message via the Server.
+//
+// In steady state, i.e. after the handshake is complete, the RemoteNode can be used to send a message to the peer,
+// retrieve the peer's handshake metadata, and close the connection with the peer. The RemoteNode has a single-use
+// lifecycle. Once the RemoteNode is terminated, it will be disposed of, and a new RemoteNode must be created if we
+// wish to reconnect to the peer in the future.
+type RemoteNode struct {
+	mtx sync.Mutex
+
+	peer *Peer
+	// The id is the unique identifier of this RemoteNode. For outbound connections, the id will be the same as the
+	// attemptId of the OutboundConnectionAttempt, and the subsequent id of the outbound peer. For inbound connections,
+	// the id will be the same as the inbound peer's id.
+	id               RemoteNodeId
+	connectionStatus RemoteNodeStatus
+
+	params *DeSoParams
+	srv    *Server
+	bc     *Blockchain
+	cmgr   *ConnectionManager
+
+	// minTxFeeRateNanosPerKB is the minimum transaction fee rate in nanos per KB that our node will accept.
+	minTxFeeRateNanosPerKB uint64
+	hyperSync              bool
+	archivalMode           bool
+	posValidator           bool
+
+	// handshakeMetadata is used to store the information received from the peer during the handshake.
+	handshakeMetadata *HandshakeMetadata
+	// keystore is a reference to the node's BLS private key storage. In the context of a RemoteNode, the keystore is
+	// used in the Verack message for validator nodes to prove ownership of the validator BLS public key.
+	keystore *BLSKeystore
+
+	// versionTimeExpected is the latest time by which we expect to receive a Version message from the peer.
+	// If the Version message is not received by this time, the connection will be terminated.
+	versionTimeExpected time.Time
+	// verackTimeExpected is the latest time by which we expect to receive a Verack message from the peer.
+	// If the Verack message is not received by this time, the connection will be terminated.
+	verackTimeExpected time.Time
+}
+
+// HandshakeMetadata stores the information received from the peer during the Version and Verack exchange.
+type HandshakeMetadata struct {
+	// ### The following fields are populated during the MsgDeSoVersion exchange.
+	// versionNonceSent is the nonce sent in the Version message to the peer.
+	versionNonceSent uint64
+	// versionNonceReceived is the nonce received in the Version message from the peer.
+	versionNonceReceived uint64
+	// userAgent is a meta level label that can be used to analyze the network.
+	userAgent string
+	// ServiceFlag is a bitfield that indicates the services supported by the peer.
+	ServiceFlag ServiceFlag
+	// StartingBlockHeight is the block height of the peer's block tip during the Version exchange.
+	StartingBlockHeight uint32
```

Comment:
We sometimes use uint64s instead. It's kinda annoying. we should probably make this a uint64 to be forward compatible as I think eventually we'll want to move all block height representations to be a uint64.